crypto/err/err.c

```diff
@@ -895,3 +896,48 @@ void err_clear_last_constant_time(int clear)
                                      0, ERR_FLAG_CLEAR);
     es->err_flags[top] |= clear;
 }
+
+static void *raise_malloc_failure(const char *file, int line, const char *func)
+{
+    /*
+     * Mem alloc error loop while reporting malloc error is prevented by using
+     * CRYPTO_zalloc() rather than OPENSSL_zalloc for ERR_STATE allocation-
+     */
+    ERR_new();
+    ERR_set_debug(file, line, func);
+    ERR_set_error(ERR_LIB_CRYPTO, ERR_R_MALLOC_FAILURE, NULL);
+    return NULL;
+}
+
+void *CRYPTO_malloc_with_err(void *(*fn)(size_t, const char *, int), size_t num,
+                             const char *file, int line, const char *func)
+{
+    void *ptr = (*fn)(num, file, line);
```

Comment:
I'm not a fan of an indirect call here.  We already do one to get to malloc et al and this makes it two.  Moreover, this one won't be predicted at all well because so many functions are funneling through here.

The alternative would be having each of the allocation functions separately defined which is ugly but won't hit performance.

Reply:
Sigh, a smart compiler might flatten this out, but likely we cannot hope for that.
Would be so nice to use just a macro for `CRYPTO_malloc_with_err` et al., but this is made impossible by the non-linear use of the `num` argument.

Reply:
Anyway, I do not really see the very slight decrease in performance using the function pointer as an issue here.
Memory allocation is in itself a pretty expensive operation, so an extra function pointer dereference should not make any noticeable difference.
Moreover, memory allocation is very unusual in tight loops.

Reply:
Could we make the functions static inline? Not sure if that doesn't open another can of worms though.

Reply:
I'm uncertain if it is possible to declare  `CRYPTO_malloc_with_err` et al. `static` because they need to be declared in a public header (`crypto.h`) for use with public macros such as `OPENSSL_malloc()`.
I was also considering making `CRYPTO_malloc_with_err` et al. `inline`, but as there is s pretty large number of calls to `OPENSSL_malloc()` et al., this would likely add quite a bit to the code size, while the performance gain would be very small.
So also in this direction I see no point trying to optimize on performance.

Reply:
> Anyway, I do not really see the very slight decrease in performance using the function pointer as an issue here. ... Moreover, memory allocation is very unusual in tight loops.

This seems inconsistent with your previous comments (don't have the PR's handy, sorry) about the cost of decoding objects. And perhaps previously-expressed concerns with overall performance.

My suggestion: add a build-time option to use backtrace and solve it for the major platform. Other platforms can be caught up as time permits. Using abort() is a generally portable way to get a backtrace, too :)

Reply:
> add a build-time option to use backtrace 

Do you mean this (which got deprecated with 3.0)?

https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/Configure#L363

Reply:
> > Anyway, I do not really see the very slight decrease in performance using the function pointer as an issue here. ... Moreover, memory allocation is very unusual in tight loops.
> 
> This seems inconsistent with your previous comments (don't have the PR's handy, sorry) about the cost of decoding objects. And perhaps previously-expressed concerns with overall performance.

Here we are not talking about decoding objects and the very high total number of malloc calls, which is quite a different story.
For this PR I do not really have performance concerns, and all I'm saying about it (in response to @paulidale's concern) is that an extra function pointer indirection is negligible as compared to the overall cost per malloc call.

> My suggestion: add a build-time option to use backtrace and solve it for the major platform. Other platforms can be caught up as time permits. Using abort() is a generally portable way to get a backtrace, too :)

This looks like a good pragmatic solution to the general debugging issue that @levitte mentioned above.
I'm in favor of such a thing, but this should be discussed and implemented independently of this PR.

Reply:
Unfortunately, last time I looked, that backtrace output was less than satisfying 😕

Reply:
@richsalz
> @DDvO, I am not sure this PR should go forward until the backtrace issue is resolved, since lots of the "extra suff" that's being added here is to capture caller info.

No, this PR is does not aim to solve the general caller info/backtrace issue.
As one can see from its title, the OP, and the subsequent discussion,
this PR 'just' aims to relieve callers of the OpenSSL malloc functions to add `ERR_raise(ERR_LIB_FOO, ERR_R_MALLOC_FAILURE)` calls on malloc failure, which was so far done more or less consistently, and sometimes even wrongly/misleadingly.
And as agreed above, existing such calls are going to be either removed (in case they directly pertain to an OpenSSL malloc call) or are retained in the generalized form `ERR_raise(ERR_LIB_FOO, ERR_R_CRYPTO_LIB)`.

Reply:
It has a spurious `state` to `s` change, that's not a big deal. It changes `C

It adds new `xxx_with_err` functions that might not be needed if the decision is to use backtrace or its equivalent. (And if those functions are only for internal use, why are they in public header files and documentation?) It does no work to change the source base to use those `with_err` functions.

This seems like a partial change, no promise of future work to complete it and a known pending open issue about the direction in which to go.

Reply:
@richsalz occasionally it feels like you wanna re-enter the OpenSSL project?

Reply:
malloc shouldn't be a huge expense most of the time.  Modern libc's implement it pretty well.  A branch misprediction takes hundreds of cycles which represents quite an expense.

The other concern is about what we make public.  Once public, forever public (almost).  Do we want `CRYPTO_malloc_with_err()` as a public API?

Reply:
Thinking a little more about this, I found a solution that does not use function pointers and the like and
avoids adding (actually internal) aux functions to public headers. It even does not change public headers at all.

It reports malloc errors once and for all in `CRYPTO_malloc()`
while making sure to prevent mem alloc error loops simply as follows:
`ossl_err_get_state_int()` in `err.c` uses the special call `CRYPTO_zalloc(num, NULL, 0)` for `ERR_STATE` allocation,
and `CRYPTO_malloc()` does not report an malloc failure in this special case `file == NULL && line == 0`.

Reply:
Nice.